Sources/performance-worker/WorkerService.swift

```diff
@@ -87,7 +89,41 @@ final class WorkerService: Grpc_Testing_WorkerService.ServiceProtocol, Sendable
   ) async throws
     -> GRPCCore.ServerResponse.Stream<Grpc_Testing_WorkerService.Method.RunServer.Output>
   {
-    throw RPCError(code: .unimplemented, message: "This RPC has not been implemented yet.")
+    return ServerResponse.Stream { writer in
+      for try await message in request.messages {
+        switch message.argtype {
+        case let .some(.setup(serverConfig)):
+          try await self.setupServer(serverConfig)
+          try await withThrowingTaskGroup(of: Void.self) { group in
+            group.addTask {
+              let role = self.state.withLockedValue({ $0.role })
+              switch role {
+              case let .server(server):
+                try await server.run()
+              default:
+                ()
+              }
+            }
+            try await group.next()
+            group.cancelAll()
+          }
+        case let .some(.mark(mark)):
+          let response = try await self.makeStatsResponse(for: mark)
+          try await writer.write(response)
+        case .none:
+          ()
+        }
+      }
+      self.state.withLockedValue {
+        switch $0.role {
+        case .server(let server):
+          server.stopListening()
+        default:
+          ()
+        }
+      }
```

Comment:
Generally speaking avoid calling out to objects while holding a lock as you don't always know what they may do (and it simply might not be necessary to hold the lock for that long). As an example, the server could have a mechanism which notifies us when the server has stopped listening, and we may want to acquire the lock in that case to update our state. If we're still holding the lock (and it isn't a re-entrant lock) then we're stuffed.

In this case we only need the lock to grab the reference to the server so `stopListening` can be called after releasing the lock.